src/libtls/tls_peer.c

```diff
@@ -1397,12 +1397,27 @@ static status_t send_certificate(private_tls_peer_t *this,
 	auth_rule_t rule;
 	bio_writer_t *certs;
 	chunk_t data;
+	tls_version_t version_min, version_max;
 
-	this->private = find_private_key(this);
+	version_min = this->tls->get_version_min(this->tls);
+	version_max = this->tls->get_version_max(this->tls);
+	if (version_max > TLS_1_1) {
+		array_t *key_types;
+
+		enumerator = tls_get_supported_key_types(version_min, version_max);
+		key_types = tls_create_common_key_types(enumerator, this->hashsig);
+		enumerator->destroy(enumerator);
+		this->private = tls_find_private_key(this->peer, this->peer_auth,
+											 key_types, NULL);
```

Comment:
As discussed, it might be possible to combine these into a single public helper method in `tls_crypto.c` (or perhaps two, as the server calls `tls_find_private_key()` multiple times with the array of key types, but even one should be possible if it's an enumerator that returns private keys and internally keeps track of the key types via array).

Reply:
I’m not quite sure how to keep track on the key types internally via an array. I think I have to do the while-loop on the array of `key_types` with `key = lib->credmgr->get_private()` directly in `tls_find_private_key()`. I could then iterate over all possible and found keys and store them in a new array. The function `tls_find_private_key()` could then return an enumerator via `array_create_enumerator()`. The caller then has to iterate over all keys but picks the first and removes it from the array with `array_remove_at()` in case it has to iterate over possible keys again later. Is this the correct approach?

Reply:
You'd actually replace `tls_find_private_key()` and make that the main loop of the enumerator, which returns TRUE if another key type (from the internally stored array) and private key (via `get_private()`) was found and FALSE if there are no remaining key types. A caller can simply call `enumerate()` to retrieve another private key, if any.